pkg/util/common/common_test.go

```diff
@@ -93,40 +93,133 @@ func TestStructToMap(t *testing.T) {
+	var defaultStruct Top
+	var nilPointer *Top = nil
+
+	cases := []struct {
+		testName string
+		input    interface{}
+		output   map[string]interface{}
+	}{
+		{
+			testName: "nil interface",
+			input:    nil,
+			output:   map[string]interface{}{},
+		},
+		{
+			testName: "integer",
+			input:    1,
+			output:   map[string]interface{}{},
+		},
+		{
+			testName: "string",
+			input:    "a",
+			output:   map[string]interface{}{},
+		},
+		{
+			testName: "default struct",
+			input:    defaultStruct,
+			output:   map[string]interface{}{},
+		},
+		{
+			testName: "valid struct",
+			input: Top{
+				Name:      "top",
+				Value:     0,
+				NestedPtr: &nested,
+				ID:        "top1",
+				MyMap: map[string]Nested{
+					"n1": Nested{
+						Foo: []MoreNested{
+							{
+								Name:         "ms1",
+								Value:        1,
+								ID:           &str,
+								JSONLessStr:  "42",
+								privateValue: 1000,
+							},
+							{
+								Name:  "ms2",
+								Value: 2,
+								ID:    nil,
+							},
+						},
+					},
 				},
 			},
-		},
-		"mymap": map[string]interface{}{
-			"n1": map[string]interface{}{
-				"moreNested": []interface{}{
-					map[string]interface{}{
-						"id":          "toto",
-						"name":        "ms1",
-						"value":       float64(1),
-						"JSONLessStr": "42",
+			output: map[string]interface{}{
+				"name":  "top",
+				"value": 0,
+				"nested": map[string]interface{}{
+					"moreNested": []interface{}{
+						map[string]interface{}{
+							"id":          "toto",
+							"name":        "ms1",
+							"value":       float64(1),
+							"JSONLessStr": "42",
+						},
+						map[string]interface{}{
+							"name":        "ms2",
+							"value":       float64(2),
+							"JSONLessStr": "",
+						},
 					},
-					map[string]interface{}{
-						"name":        "ms2",
-						"value":       float64(2),
-						"JSONLessStr": "",
+				},
+				"mymap": map[string]interface{}{
+					"n1": map[string]interface{}{
+						"moreNested": []interface{}{
+							map[string]interface{}{
+								"id":          "toto",
+								"name":        "ms1",
+								"value":       float64(1),
+								"JSONLessStr": "42",
+							},
+							map[string]interface{}{
+								"name":        "ms2",
+								"value":       float64(2),
+								"JSONLessStr": "",
+							},
+						},
 					},
 				},
 			},
 		},
-	}, StructToMap(top))
+		{
+			testName: "slice of structs",
+			input: []Top{
+				Top{
+					Name:      "first",
+					Value:     0,
+					NestedPtr: nil,
+					ID:        "first1",
+					MyMap:     map[string]Nested{},
+				},
+				Top{
+					Name:      "second",
+					Value:     0,
+					NestedPtr: nil,
+					ID:        "second2",
+					MyMap:     map[string]Nested{},
+				},
+			},
+			output: map[string]interface{}{},
+		},
+		{
+			testName: "nil pointer",
+			input:    nilPointer,
+			output:   map[string]interface{}{},
+		},
+		{
+			testName: "pointer",
+			input:    &top,
+			output:   map[string]interface{}{},
```

Comment:
The expected output for a pointer to a valid struct is proposed to be the same as the output for that struct.  Just not sure currently how to access the entire struct from a pointer, running into` invalid indirect of obj` errors when trying to use `*obj`.  Do I need to rebuild it or is there any other method similar to unmarshaling that can be used?

Reply:
done with: 
```
	if rt.Kind() == reflect.Ptr {
		obj = rv.Elem().Interface()
		rt, rv = reflect.TypeOf(obj), reflect.ValueOf(obj)
	}
```